src/backups.py

```diff
@@ -0,0 +1,301 @@
+
+        # Retrieve and validate missing S3 parameters
+        s3_parameters, missing_parameters = self._retrieve_s3_parameters()
+        if missing_parameters:
+            event.fail(f"Missing S3 parameters: {missing_parameters}")
+            return
+
+        s3_directory = f"{s3_parameters['path']}/{datetime_backup_requested}"
+
+        # Check if this unit can perform backup
+        can_unit_perform_backup, validation_message = self._can_unit_perform_backup()
+        if not can_unit_perform_backup:
+            logger.warning(validation_message)
+            event.fail(validation_message)
+            return
+
+        # Test uploading metadata to S3 to test credentials before backup
+        juju_version = JujuVersion.from_environ()
+        metadata = f"""Date Backup Requested: {datetime_backup_requested}
+Model Name: {self.model.name}
+Application Name: {self.model.app.name}
+Unit Name: {self.charm.unit.name}
+Juju Version: {str(juju_version)}
+"""
+        success = upload_content_to_s3(
+            metadata,
+            s3_parameters["bucket"],
+            f"{s3_directory}/metadata",
+            s3_parameters["region"],
+            s3_parameters["endpoint"],
+            s3_parameters["access-key"],
+            s3_parameters["secret-key"],
+        )
+        if not success:
+            event.fail("Failed to upload metadata to provided S3")
+            return
+
+        # Run operations to prepare for the backup
+        success, error_message = self._pre_backup()
+        if not success:
+            logger.warning(error_message)
+            event.fail(error_message)
+            return
+
+        # Perform the backup
+        success, error_message = self._backup(
+            s3_parameters["bucket"],
+            s3_directory,
+            s3_parameters["endpoint"],
+            s3_parameters["region"],
+            s3_parameters["access-key"],
+            s3_parameters["secret-key"],
+        )
+        if not success:
+            logger.warning(error_message)
+            event.fail(error_message)
+
+            success, error_message = self._post_backup()
+            if not success:
+                logger.warning(error_message)
+
+            return
+
+        # Run operations to clean up after the backup
+        success, error_message = self._post_backup()
+        if not success:
+            logger.warning(error_message)
+            event.fail(error_message)
+            return
+
+        event.set_results(
+            {
+                "backup-id": datetime_backup_requested,
+            }
+        )
+
+    def _retrieve_s3_parameters(self) -> Tuple[Dict, List[str]]:
+        """Retrieve S3 parameters from the S3 integrator relation."""
+        s3_parameters = self.s3_integrator.get_s3_connection_info()
+
+        required_parameters = [
+            "bucket",
+            "endpoint",
+            "region",
+            "path",
+            "access-key",
+            "secret-key",
+        ]
```

Comment:
only `access-key`, `secret-key` and `bucket` are required, all other fields are optional. S3 can work without them.

Reply:
addressed in aedeff6b9921c57431722fcf5a827a4509226ae3. made the remaining s3 parameters optional